shared/src/types/UtilityTypes.ts

```diff
@@ -0,0 +1,9 @@
+// credits goes to https://stackoverflow.com/a/50375286
+// function intersection producec - functin overloads
+export type UnionToIntersection<U> = (
```

Comment:
🤯

Reply:
Hahaha don't get me started on this 😂 But basically it was a deep ended hole getting a type object structure:
```javascript
 { eventName: { eventProp: type } }
```
Into be type checked in a function where they are separate arguments i.e. `logEvent(eventName, { eventProp: type });`.
What it does in the end is to create multiple function "overloads" of every possible combination:
```typescript
('some event name', { 'some event prop': string}) => void &
('some second event name', { 'some other event prop': boolean }) => void &
('some third event name', { 'some other event prop': boolean }) => void &
```
... and so forth...

Does it make sense? :D